acceptance/github_asset_fetcher.go

```diff
@@ -0,0 +1,420 @@
+		return destPath, nil
+	}
+
+	release, _, err := f.githubClient.Repositories.GetReleaseByTag(f.ctx, owner, repo, version)
+	if err != nil {
+		return "", errors.Wrap(err, "getting release")
+	}
+
+	var desiredAsset *github.ReleaseAsset
+	for _, asset := range release.Assets {
+		if expr.MatchString(*asset.Name) {
+			desiredAsset = asset
+			break
+		}
+	}
+	if desiredAsset == nil {
+		return "", fmt.Errorf("could not find asset matching expression %s", expr.String())
+	}
+
+	var returnPath string
+	if extract && strings.Contains(*desiredAsset.Name, ".tgz") {
+		targetDir := filepath.Join(f.cacheDir, stripExtension(*desiredAsset.Name))
+		if err := os.MkdirAll(targetDir, 0755); err != nil {
+			return "", err
+		}
+
+		if err := f.downloadAndExtractTgz(*desiredAsset.BrowserDownloadURL, targetDir); err != nil {
+			return "", err
+		}
+
+		returnPath = targetDir
+	} else if extract && strings.Contains(*desiredAsset.Name, ".zip") {
+		targetPath := filepath.Join(f.cacheDir, *desiredAsset.Name)
+		if err := f.downloadAndExtractZip(*desiredAsset.BrowserDownloadURL, targetPath); err != nil {
+			return "", err
+		}
+
+		returnPath = stripExtension(targetPath)
+	} else {
+		targetPath := filepath.Join(f.cacheDir, *desiredAsset.Name)
+		err = f.downloadAndSave(*desiredAsset.BrowserDownloadURL, targetPath)
+		if err != nil {
+			return "", err
+		}
+
+		returnPath = targetPath
+	}
+
+	err = f.writeCacheManifest(owner, repo, func(cache assetCache) {
+		existingAssets, found := cache[owner][repo].Assets[version]
+		if found {
+			cache[owner][repo].Assets[version] = append(existingAssets, returnPath)
+		}
+		cache[owner][repo].Assets[version] = []string{returnPath}
+	})
+	if err != nil {
+		f.logger.Warn(errors.Wrap(err, "failed to write cache").Error())
+	}
+	return returnPath, nil
+}
+
+func (f *GithubAssetFetcher) FetchReleaseSource(owner, repo, version string) (string, error) {
+	if destDir, _ := f.cachedSource(owner, repo, version); destDir != "" {
+		fmt.Printf("found %s in cache for %s/%s %s\n", destDir, owner, repo, version)
+		return destDir, nil
+	}
+
+	release, _, err := f.githubClient.Repositories.GetReleaseByTag(f.ctx, owner, repo, version)
+	if err != nil {
+		return "", errors.Wrap(err, "getting release")
+	}
+
+	destDir := filepath.Join(f.cacheDir, strings.ReplaceAll(*release.Name, " ", "-")+"-source")
+	if err := os.MkdirAll(destDir, 0755); err != nil {
+		return "", err
+	}
+
+	if err := f.downloadAndExtractTgz(*release.TarballURL, destDir); err != nil {
+		return "", err
+	}
+
+	err = f.writeCacheManifest(owner, repo, func(cache assetCache) {
+		cache[owner][repo].Sources[version] = destDir
+	})
+	if err != nil {
+		f.logger.Warn(errors.Wrap(err, "failed to write cache").Error())
+	}
+	return destDir, nil
+}
+
+func (f *GithubAssetFetcher) FetchReleaseVersion(owner, repo string, n int) (string, error) {
+	if version, _ := f.cachedVersion(owner, repo, n); version != "" {
```

Comment:
Cache the version as well to avoid rate limiting. In theory if we blow away the cache manifest but the same release versions exist, we just need to query the API once to get the versions and can re-use the assets that would still be in the cache directory.